mewarpx/mewarpx/assemblies.py

```diff
@@ -294,3 +299,194 @@ def _install_in_simulation(self):
+        transformed_points = self._transform_coordinates(points)
+        idx = self._sort_particles(transformed_points)
+        normals = self.region_normals[idx]
+        nhat = np.zeros((3, (normals.shape[0])))
+        nhat[0, :] = normals[:, 0]
+        nhat[2, :] = normals[:, 1]
+        return nhat
+
+    def _get_transformation_matrix(self):
+        """Tranform1 is used to move the rectangle to the appropriate location
+        in the plane and scale it to the appropriate size. The rectangle is
+        placed such that the longest side is along the z-axis and the bottom
+        right corner is at (1, -1). This placement makes it simple to pick out
+        positions that are in regions 0 or 1 as well as positions in region 4
+        or 5.
+        """
+        # first transpose the box to the origin
+        transpose1 = np.identity(3)
+        transpose1[0, 2] = (-self.center_x + self.length_x / 2.0)
+        transpose1[1, 2] = (-self.center_z + self.length_z / 2.0)
+
+        # next scale the axes to make zone 0 equilateral with side length 1
+        l = min(self.length_x, self.length_z)
+        scale = np.identity(3)
+        scale[0, 0] = 2.0 / l
+        scale[1, 1] = 2.0 / l
+
+        # next transpose again so the center is at the origin
+        transpose2 = np.identity(3)
+        transpose2[0, 2] = -1
+        transpose2[1, 2] = -1
+
+        # rotate by 90 deg if the length_z is smallest
+        rotate = np.identity(3)
+        if self.length_z < self.length_x:
+            rotate[0,0] = np.cos(np.pi/2)
+            rotate[0,1] = -np.sin(np.pi/2)
+            rotate[1,0] = np.sin(np.pi/2)
+            rotate[1,1] = np.cos(np.pi/2)
+
+        transformation_matrix = np.dot(
+            np.dot(rotate, transpose2),
+            np.dot(scale, transpose1)
+        )
+        return transformation_matrix
+
+    def _transform_coordinates(self, points_in):
+        """Transforms the coordinates of the particles using the
+        transformation matrix of the rectangle"""
+        points = np.ones((len(points_in), 3))
+        points[:, 0] = points_in[:, 0]
+        points[:, 1] = points_in[:, 1]
+
+        points = points.dot(self.transformation_matrix.T)
+
+        return points[:, 0:2]
+
+    def _sort_particles(self, points):
+        """Finds the regions of the rectangle that each point belongs to"""
+        # the checks below isolate region in which the given position cannot
+        # be, for example if c0 is True for a specific position it means
+        # that position cannot be in region 0.
+        c0 = np.array(points[:, 1] > -abs(points[:, 0]), dtype=int)
+        c3 = np.array(points[:, 1] - self.scaled_h + 2.0 < abs(points[:, 0]), dtype=int)
+        c1 = np.array(points[:, 0] <= 0, dtype=int)
+
+        # the booleans below form bit values to build up the index of the
+        # region in which a given position lies.
+        b1 = np.logical_not(c3) | (c1 & c0)
+        b2 = c0 & c3
+        idx = 2*b1 + b2
```

Comment:
Please check this. It should be logically identical if I did it correctly. But your c0 and c3 (which I think should have been c2) relied on being a definitive in/out, eg if c0 was False the particle was definitely in 0 (whereas if c1 was False, particle still wasn't necessarily in 1). So it seemed better to me to negate those to indicate inclusion rather than exclusion. Admittedly I have 5 boolean operations rather than 4...
```suggestion
        # The checks below isolate the region in which the given position lies.
        # c0 and c2 are True iff a particle is in region 0 or 2, respectively
        # not_c1 indicates a particle may be in regions 0, 2, or 3. 
        c0 = np.array(points[:, 1] < -abs(points[:, 0]), dtype=int)
        c2 = np.array(points[:, 1] - self.scaled_h + 2.0 > abs(points[:, 0]), dtype=int)
        not_c1 = np.array(points[:, 0] <= 0, dtype=int)

        # the booleans below form bit values to build up the index of the
        # region in which a given position lies.
        # b1 is True if in region 2 or 3.
        b1 = c2 | (not_c1 & np.logical_not(c0))
        # b2 is True if in region 1 or 3; equivalently, if it's not in 0 or 2.
        b2 = np.logical_not(c0 | c2)
        idx = 2*b1 + b2
```

Reply:
Actually I take that back, I think the `c2 |` in `b1` is actually superfluous which would bring you back down to 4 bools. If you're not in c0 and not in c1 you're definitely in 2 or 3...

Reply:
This makes sense to me! I tested it and it works too. I'll put the change in, but I also want to change the naming of b1 and b2, since the bit numbering isn't correct.